Mode picker persistence

julesian opened the pull request:
Added mode picker persistence. 

Added `allModes` reference to `filteredNearby` to determine which modes are disabled

mode distinction is determined via `identifier` and `localImageName` as I wanted to reduce saving too much data to UserDefaults. As I see that the distinction of TKModeInfo is via its variables (Equatable). 

Would be nice if there is a single variable that uniquely identifies the mode since `identifier` seems to be showing a group-type value, and gets every beam / neuron modes alike.

On other note: @nighthawk `identifier` seems to be a group-type value which is good for sorting initially #186 but there seems to be grouping based on location, so that might be an issue. I get now why you mentioned before that this is not a good group identifier.

nighthawk commented on Sources/TripKit/helpers/TKUserProfileHelper.swift:
```diff
@@ -132,5 +133,73 @@ public class TKUserProfileHelper: NSObject {
+  }
+  
+  // combined both identifier and imageName to identify the actual mode, since identifier is the same for both bike and scooter modes
+  @objc public class func identifier(from mode: TKModeInfo) -> String? {
```
Might be better to keep this one private, or is there a case where this need to be accessible?

```suggestion
  @objc private class func identifier(from mode: TKModeInfo) -> String? {
```

julesian replied:
ah yes, some of the methods here were previously used outside and decided to have it only in the class file

nighthawk commented on Sources/TripKit/helpers/TKUserProfileHelper.swift:
```diff
@@ -132,5 +133,73 @@ public class TKUserProfileHelper: NSObject {
   
+  // MARK: - Mode by mode, Mode picker
+  
+  @objc public class func isSharedVehicleModeEnabled(identifier: ModeIdentifier) -> Bool {
+    return disabledSharedVehicleModes.contains(identifier)
+  }
+  
```
Would be best to remove this function as just have the public interface based on `TKModeInfo` for simplicity. 

Also, if you added a regular mode identifier to `disabledSharedVehicleModes`, then calling `isSharedVehicleModeEnabled(mode:)` can crash if there's no ":" in the identifier and therefore no `components[1]`.

```suggestion
```

julesian replied:
oh okay, noted on this one.

nighthawk commented on Sources/TripKit/helpers/TKUserProfileHelper.swift:
```diff
@@ -132,5 +133,73 @@ public class TKUserProfileHelper: NSObject {
+  @objc public class func update(pickedModes: Set<TKModeInfo>, allModes: Set<TKModeInfo>) {
+    let pickedIds = pickedModes.compactMap { TKUserProfileHelper.identifier(from: $0) }
+    let allIds = allModes.compactMap { TKUserProfileHelper.identifier(from: $0) }
+    update(pickedIds: pickedIds, allIds: allIds)
+  }
+  
+  @objc public class func update(pickedIds: [ModeIdentifier], allIds: [ModeIdentifier]) {
```
As above, let's remove the `updated(pickedIds:allIds)` variant.

```suggestion
```

nighthawk commented on Sources/TripKit/helpers/TKUserProfileHelper.swift:
```diff
@@ -132,5 +133,73 @@ public class TKUserProfileHelper: NSObject {
+    return disabledSharedVehicleModes.count > 0
+  }
+    
+  @objc public class var disabledSharedVehicleModes: [ModeIdentifier] {
```
```suggestion
  private class var disabledSharedVehicleModes: [ModeIdentifier] {
```

nighthawk reviewed (changes requested):
Couple of suggests to simplify the public interface and avoid a potential crash.

julesian commented:
updated, made saving to user defaults raw data instead of a partial combined string to identify `TKModeInfo`, which follows `isEqual`. A bit eerie on saving type Data to UserDefaults but it is small in size.

Removed some of the methods.

Decided not to decode values in `disabledSharedVehicleModes` and compare directly from newly encoded values to save performance.

nighthawk commented on Sources/TripKit/model/TKModeInfo.swift:
```diff
@@ -49,6 +49,23 @@ public class TKModeInfo: NSObject, Codable, NSSecureCoding {
+    else {
+      return true
+    }
+    var hasDisabled = false
+    for mode in disabledSharedVehicleModes {
+      if encoded == mode {
+        hasDisabled = true
+        break
+      }
+    }
+    return !hasDisabled
```
This can be simplified to something like:

```suggestion
    return !disabledSharedVehicleModes.contains(encoded)
```

julesian replied:
ah yes.

nighthawk commented on Sources/TripKit/helpers/TKUserProfileHelper.swift:
```diff
@@ -132,5 +133,40 @@ public class TKUserProfileHelper: NSObject {
+    UserDefaults.shared.set(modes, forKey: DefaultsKey.disabled.rawValue)
+  }
+    
+  @objc public class var disabledSharedVehicleModes: [Data] {
```
Best to make this internal to keep the "Data" part hidden from the public interface:

```suggestion
  class var disabledSharedVehicleModes: [Data] {
```

julesian replied:
I see, noted.

nighthawk commented on Sources/TripKit/helpers/TKUserProfileHelper.swift:
```diff
@@ -132,5 +133,40 @@ public class TKUserProfileHelper: NSObject {
+      !pickedIds.contains($0)
+    }
+    
+    pickedIds.forEach { mode in
+      if modes.contains(mode) {
+        modes = modes.filter { mode != $0 }
+      }
+    }
```
I find this little snippet hard to read. What this does is remove from `modes` everything that's in `pickedIds`, right? 

That's what `mutating func subtract(_ other: [Set])` does. So this can be just:

```suggestion
    modes.subtract(pickedIds)
```

julesian replied:
that is correct. thats a handy function.

nighthawk commented on Sources/TripKit/helpers/TKUserProfileHelper.swift:
```diff
@@ -132,5 +133,40 @@ public class TKUserProfileHelper: NSObject {
+      }
+    }
+    
+    toDisable.forEach {
+      if !modes.contains($0) {
+        modes.append($0)
+      }
+    }
```
This is the union of `toDisable` and `modes`. So it can be just:

```suggestion
    modes.formUnion(toDisable)
```

nighthawk reviewed (commented):
Some stylistic suggestions

julesian commented:
updated

nighthawk reviewed (approved):
Thank you. LGTM.
